cmd/mistryd/server.go

```diff
@@ -118,28 +126,82 @@ func (s *Server) HandleNewJob(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// send the work item to the worker pool
+	workItem := WorkItem{j, jr, make(chan WorkResult, 1)}
+
+	select {
+	case s.workQueue <- workItem:
+	default:
+		// the in-memory queue is overloaded, we have to wait for the workers to pick
+		// up new items.
+		// return a 503 to signal that the server is overloaded and for clients to try
+		// again later
+		// 503 is an appropriate status code to signal that the server is overloaded
+		// for all users, while 429 would have been used if we implemented user-specific
+		// throttling
+		s.Log.Print("Failed to send message to work queue")
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
+
+	// if async, we're done, otherwise wait for the result in the result channel
 	if _, isAsync := r.URL.Query()["async"]; isAsync {
-		s.handleNewJobAsync(j, jr, w)
+		s.Log.Printf("Scheduled %s...", j)
+		w.WriteHeader(http.StatusCreated)
 	} else {
-		s.handleNewJobSync(j, jr, w)
+		s.Log.Printf("Waiting for result for %s...", j)
+		result, ok := <-workItem.ResultChan
+		if !ok {
+			s.Log.Print("Failed to read from result channel")
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+		s.Log.Printf("Received result for %s, returning", j)
+		s.writeJobResultResponse(j, result, w)
 	}
 }
 
-// handleNewJobSync triggers the build synchronously, and writes the
-// build result JSON to the response
-func (s *Server) handleNewJobSync(j *Job, jr types.JobRequest, w http.ResponseWriter) {
-	s.Log.Printf("Building %s...", j)
-	buildInfo, err := s.Work(context.Background(), j, jr)
-	if err != nil {
-		http.Error(w, fmt.Sprintf("Error building %s: %s", j, err),
+// WorkResult contains the result of a job, either a buildinfo or an error
+type WorkResult struct {
+	BuildInfo *types.BuildInfo
+	Err       error
+}
+
+// WorkItem contains a job and a channel to place the job result
+type WorkItem struct {
+	Job        *Job
+	JobRequest types.JobRequest
+	ResultChan chan WorkResult
+}
+
+func (s *Server) worker(id int) {
```

Comment:
Needs a more accurate name (work?) and documentation. Also, do we need to have an `id`?